Add numerical ordering option for string comparison operations

PranavSenthilnathan opened the pull request:
Adds numerical ordering for comparison operations (e.g. Compare, Equals, GetHashCode, GetSortKey). This now enables comparisons of numbers based on their numerical value instead of lexicographical order, such as 2 < 10. We don't support Index operations (e.g. StartWith, EndsWith, IsPrefix, IsSuffix) since the underlying globalization libraries, NLS and ICU, don't support it.

Because this new option relies on underlying globalization libraries, there could be differences in behavior for different platforms and libraries:
- Full ICU (default in most cases):
  - Numbers with leading zeros are treated as equal to equivalent numbers with no leading zeros (`"01" == "1"`).
  - Equivalent numbers in different languages/scripts are treated the same (`"1" == "١"`, where ١ is the Arabic-Indic Digit One).
- NLS (legacy Windows):
  - Numbers with leading zeros are treated as unequal to equivalent numbers with no leading zeros (`"01" != "1"`). However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.
  - Equivalent numbers in different languages/scripts are treated as unequal (`"1" != "١"`)
- Hybrid ICU:
  - WASM: The same as ICU, but the option set is limited. This PR does not change the valid option set besides allowing adding the NumericOrdering option to any previously valid option. [See this](https://github.com/dotnet/runtime/blob/f87428bdd39418a758d916bd950b82e3ee69835f/docs/design/features/globalization-hybrid-mode.md#wasm) for the allowed options.
  - iOS, tvos, maccatalyst: Not supported because I don't have an Apple device to test with. However, Apple does support numeric comparisons for strings.
    - [NSString.CompareOptions API](https://developer.apple.com/documentation/foundation/nsstring/compareoptions/1415530-numeric)
    - PAL change would probably be to recognize the dotnet CompareOptions and convert it to the NSString.CompareOptions [here](https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Globalization.Native/pal_collation.m#L70)

Closes #13979.

PranavSenthilnathan committed:
Numerical ordering for string compare/equals/hashcode

PranavSenthilnathan committed:
fix invariant tests

PranavSenthilnathan committed:
fix wasm tests

PranavSenthilnathan committed:
fix ios and nls tests

PranavSenthilnathan committed:
flipped condition

PranavSenthilnathan committed:
docs

PranavSenthilnathan committed:
Merge branch 'main' of https://github.com/dotnet/runtime into logical-str-cmp

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs:
```diff
@@ -3,17 +3,73 @@
+        /// Indicates that the string comparison must sort sequences of digits (Unicode general category "Nd") based on their numeric value.
+        /// For example, "2" comes before "10". Non-digit characters such as decimal points, minus or plus signs, etc.
+        /// are not considered as part of the sequence and will terminate it.
+        /// </summary>
```
>  /// </summary> [](http://example.com/codeflow?start=6&length=16)

will be good to add a remarks to this one giving more information like this option can be used in comparisons but not for search (IndexOf/StartsWith/EndsWith). Will be good to hint the behavior difference too when ICU is used against NLS. And last tell this option cannot be combined with the ordinal operations,

PranavSenthilnathan replied:
The comment was getting long so I just included just the part about indexing. Mentioning search might be confusing because people might consider IndexOf (which isn't supported) as search and GetHashCode (which is supported) as not. I think it's easier to just say that NumericOrdering works in all cases except for indexing.

I prefer to keep the combination behavior with Ordinal and OrdinalIgnoreCase on those members since this is really a property of theirs instead of numeric ordering. This is also consistent with the other options as well which don't mention Ordinal even though they can't combine with them either.

I think the ICU and NLS differences should probably go in docs rather than in doc comments since NLS usage is not going to be high. There are already docs about NLS/ICU differences that we can append to (https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization-icu#behavioral-differences).

tarekgh replied:
Well, if you prefer that, it will be better to edit the docs of the indexing APIs and add the remark there. I am trying to make it easy for the API users to understand when this new enum value is not allowed. I guess users can be puzzled if they get exceptions and do not understand what is wrong. 

You don't have to block the PR on that, but it will be good to open a doc issue/PR to add the info as needed.

PranavSenthilnathan replied:
I like that idea. I just checked and we already list the valid options so I will update those docs once this goes in:
[Compare](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.compare?view=net-8.0#system-globalization-compareinfo-compare(system-string-system-string-system-globalization-compareoptions))
[IndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.indexof?view=net-8.0#system-globalization-compareinfo-indexof(system-string-system-char-system-int32-system-globalization-compareoptions))

tarekgh commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -9186,6 +9186,7 @@ public enum CompareOptions
+        NumericOrdering = 32,
         OrdinalIgnoreCase = 268435456,
         StringSort = 536870912,
         Ordinal = 1073741824,
```
nit: should we change this to use hex numbers for the sake of the readability?

PranavSenthilnathan replied:
This is how the GetAPI tool generates it and the guidance I've heard is to make as few diffs from that tool as possible.

tarekgh replied:
Yes, but still can make diffs :-)

tannergooding replied:
The ref assembly isn’t really designed for readability, it’s designed to be correct and automatically generated

Deviations and manual diffs just cause later downstream pain and hinder the ability to rerun the tool, as people have to fight against it 

The better option would be to submit a bug or better a patch such that the output produced by the tool for flags enabled enums is “better” such as using hex or logical shifts to represent the bits instead (`1 << 0, 1 << 1, 1 << 2`, etc)

PranavSenthilnathan replied:
I've created an issue: https://github.com/dotnet/sdk/issues/44999. GenAPI seems to be in a kind of limbo state where there is a new Roslyn based version we want to switch to (tracked by https://github.com/dotnet/sdk/issues/31088) and we don't want to maintain the current CCI-based one (https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.GenAPI/README.md).

tarekgh commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/StringComparerTests.cs:
```diff
@@ -136,6 +136,27 @@ public void CreateCultureOptions_CreatesValidComparer()
             Assert.Equal(1, c.Compare("42", null));
             Assert.Throws<ArgumentException>(() => c.Compare(42, "84"));
+
+            c = StringComparer.Create(CultureInfo.InvariantCulture, CompareOptions.NumericOrdering);
```
Could you please test Equals method too here?

tarekgh reviewed (approved):
I left a few minor comments, LGTM otherwise. Thanks @PranavSenthilnathan!

I'll let @matouskozak and @ilonatommy comment on the hybrid and WASM stuff.

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

matouskozak reviewed (commented):
Thank you for considering the Apple mobile globalization :). Could you please create an issue for the missing support on iOS/..., linking the original issue and the findings that you mentioned in the PR description. I'll look into it later.

I think we might have to temporarily disable the newly added tests for Apple mobile otherwise we will start getting `PlatformNotSupportedException` for the numeric ordering tests. I started the Apple mobile CI test runs to see if that's the case.


One more thing, in the PR description, there is an example for NLS
> However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.

Does that mean that numbers with leading zeros are always smaller than without. Also, `002 < 02` then?

PranavSenthilnathan committed:
Address comments

PranavSenthilnathan commented:
> Thank you for considering the Apple mobile globalization :). Could you please create an issue for the missing support on iOS/..., linking the original issue and the findings that you mentioned in the PR description. I'll look into it later.

Created #109999

> I think we might have to temporarily disable the newly added tests for Apple mobile otherwise we will start getting `PlatformNotSupportedException` for the numeric ordering tests. I started the Apple mobile CI test runs to see if that's the case.

Updated the PR to skip them. I introduced a new PlatformDetection property IsNumericComparisonSupported instead of reusing the IsHybridGlobalizationOnApplePlatform so it's easier to find and remove once when you get the test working.

> One more thing, in the PR description, there is an example for NLS
> 
> > However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.
> 
> Does that mean that numbers with leading zeros are always smaller than without. Also, `002 < 02` then?

Yes, if the numbers are actually equal, then in NLS the one with more leading zeros is considered less (it's basically a tiebreaker). This behavior is better for deterministic sorting of lists, but the downside is that hash tables won't consider these equal. I prefer ICU's behavior here (JS/wasm does the same) but I don't think there's much we can do about it.

matouskozak commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs:
```diff
@@ -716,6 +716,11 @@ private unsafe SortKey IcuCreateSortKey(string source, CompareOptions options)
                     throw new PlatformNotSupportedException(GetPNSEWithReason("CreateSortKey", "non-invariant culture"));
                 return InvariantCreateSortKey(source, options);
             }
+#elif TARGET_MACCATALYST || TARGET_IOS || TARGET_TVOS
```
I think this `#if` should be moved after the `if ((options & ValidCompareMaskOffFlags) != 0)` exception guard or the `AssertExtensions.Throws<ArgumentException>("options", () => ci.GetSortKey(s1, CompareOptions.Ordinal));` test from `SortKeyMiscTest` should get temporarily disabled for apple mobile.

PranavSenthilnathan committed:
move platform check after argument check

PranavSenthilnathan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

matouskozak reviewed (approved):
Thank you for the test fixes. I checked the apple mobile CI and everything failing looks to be unrelated and tracked at https://github.com/dotnet/runtime/issues/103472. Looking good from Apple mobile side.

@ilonatommy do you want to check the WASM changes?

ilonatommy commented:
> @ilonatommy do you want to check the WASM changes?

Sorry, I was away for a while. I read the WASM part and it looks good. Let me just run extra platforms with `_HybridGlobalization` suffix.

ilonatommy commented:
/azp run runtime-extra-platforms

ilonatommy commented:
/azp run runtime-wasm-libtests

ilonatommy commented:
/azp run runtime-wasm

ilonatommy reviewed (approved):
`HybridGlobalization` failures are not connected.
